### PR TITLE
Личный проект: меняй-удаляй (часть 1)

### PR DESCRIPTION
**Pull request by SampetovaN**
---
:mortar_board: [Личный проект: меняй-удаляй (часть 1)](https://up.htmlacademy.ru/ecmascript/12/user/1236041/tasks/22)

### CODE REVIEW

File: src/presenter/films-board.js [hunk 1]
```diff
@@ -2,143 +2,187 @@ import FilmsListView from "./../view/films-list.js";
...
 
-    if (this._renderedFilmsCount >= this._films.length) {
+    if (this._renderedFilmsCount >= filmsCount) {
       remove(this._showMoreButtonComponent);
     }
   }
+  _renderShowMoreButton() {
+    if (this._showMoreButtonComponent !== null) {
+      this._showMoreButtonComponent = null;
+    }
```
Comment (AlexMasterov) [inline src/presenter/films-board.js:112]: https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/films-board.js#L114

<details>
<summary>_renderShowMoreButton</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/films-board.js#L110-L112
</details>

Нужно удалить всё условие с проверкой и установкой `null`.

File: src/presenter/films-board.js [hunk 2]
```diff
@@ -2,143 +2,187 @@ import FilmsListView from "./../view/films-list.js";
...
 
+    this._currentSortType = sortType;
+    this._clear({resetRenderedFilmsCount: true});
+    this._render();
+  }
   _renderSort() {
-    render(this._filmsContainer, this._sortComponent, RenderPosition.BEFORE_BEGIN);
+    if (this._sortComponent !== null) {
+      this._sortComponent = null;
+    }
```
Comment (AlexMasterov) [inline src/presenter/films-board.js:81]: https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/films-board.js#L83
<details>
<summary>_renderSort</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/films-board.js#L79-L81
</details>

Нужно удалить всё условие с проверкой и установкой `null`.

File: src/presenter/films-board.js [hunk 3]
```diff
@@ -2,143 +2,187 @@ import FilmsListView from "./../view/films-list.js";
...
-    this._filmCardPresenter[film.id] = filmCardPresenter;
-  }
 
-  _handleFilmChange(updatedFilm) {
-    this._films = updateItem(this._films, updatedFilm);
-    this._filmCardPresenter[updatedFilm.id].render(updatedFilm);
+  _getFilms() {
+    const filterType = this._filterModel.filter;
+    const films = this._filmsModel.films.slice();
+    const filteredFilms = filter[filterType](films);
```
Comment (AlexMasterov) [inline src/presenter/films-board.js:51]: https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/films-board.js#L50

`.slice` не нужен, ведь `filter` уже возвращает новый массив.

File: src/presenter/filter.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      return;
+    }
+
+    this._filterModel.setFilter(UpdateType.MAJOR, filterType);
+  }
+
+  get filters() {
+    const films = this._filmsModel.films;
+
+    return [
```
Comment (AlexMasterov) [inline src/presenter/filter.js:55]: <details>
<summary>filters</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L58
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L62
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L67
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L72
</details>

Б5~

File: src/presenter/filter.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import FilterView from "../view/filter.js";
+import {render, replace, remove} from "../utils/dom.js";
+import {filter} from "../utils/filter.js";
+import {FilterType, UpdateType} from "../utils/const.js";
+
+class Filter {
```
Comment (AlexMasterov) [inline src/presenter/filter.js:6]: <details>
<summary>Filter</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L8-L9
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L11
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/filter.js#L13
</details>

Д3

<details>
<summary>Покажи пример исправления</summary>

```diff
- this._filterModel = filterModel;
+ this._model = filterModel;
```

```diff
- this._filterComponent = null;
+ this._view = null;
```
</details>

File: src/presenter/film-card.js [hunk 6]
```diff
@@ -32,41 +34,47 @@ class FilmCard {
...
+    this._setHandlers();
     if (prevComponent === null) {
       render(this._container, this._component);
       return;
     }
     if (this._container.element.contains(prevComponent.element)) {
       replace(this._component, prevComponent);
     }
-    if (document.body.contains(prevFullComponent.element)) {
+    if (document.contains(prevFullComponent.element)) {
```
Comment (AlexMasterov) [inline src/presenter/film-card.js:45]: https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/presenter/film-card.js#L55

```diff
- if (document.contains(prevFullComponent.element)) {
+ if (document.body.contains(prevFullComponent.element)) {
```

File: src/utils/const.js [hunk 7]
```diff
@@ -21,4 +21,25 @@ const SortType = {
...
+
+const FilterType = {
+  ALL: `all`,
+  WATCHLIST: `watchlist`,
+  HISTORY: `history`,
+  FAVORITES: `favorites`
+};
+
+const COMMENT_AUTHORS = [`Tim Macoveev`, `John Doe`, `Peter Parker`, `Alice Cooper`, `Tamara Green`];
+export {TIME_MARKS, FilmListTitleOption, SortType, WatchingDateYearRange, MonthRange, DAY_DATE, UserAction, UpdateType, FilterType, COMMENT_AUTHORS};
```
Comment (AlexMasterov) [inline src/utils/const.js:45]: <details>
<summary>utils/const.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/const.js#L1-L4
</details>

Перенеси сразу в `mock/const.js`.

File: src/utils/const.js [hunk 8]
```diff
@@ -21,4 +21,25 @@ const SortType = {
...
+};
+
+const FilterType = {
+  ALL: `all`,
+  WATCHLIST: `watchlist`,
+  HISTORY: `history`,
+  FAVORITES: `favorites`
+};
+
+const COMMENT_AUTHORS = [`Tim Macoveev`, `John Doe`, `Peter Parker`, `Alice Cooper`, `Tamara Green`];
```
Comment (AlexMasterov) [inline src/utils/const.js:44]: `COMMENT_AUTHORS` нужно удалить.

File: src/main.js [hunk 9]
```diff
@@ -3,31 +3,33 @@ import StatisticsFooterView from "./view/statistics-footer.js";
...
 import FilmsView from "./view/films.js";
-import FilterView from "./view/filter.js";
 import {render} from "./utils/dom.js";
 import {generateFilmCard} from "./mock/film-card.js";
-import {generateFilter} from "./mock/filter.js";
-import {generateStatistics} from "./mock/statistics.js";
 import FilmsBoardPresenter from "./presenter/films-board.js";
+import FilmsModel from "./model/films.js";
+import FilterModel from "./model/filter.js";
+import FilterPresenter from "./presenter/filter.js";
```
Comment (AlexMasterov) [inline src/main.js:12]: <details>
<summary>utils/const.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/const.js#L5-L9
</details>

<details>
<summary>utils/utils.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/utils.js#L23-L30
</details>

Перенеси сразу в `mock/random.js` или `mock/film-card.js`.

> Скоро удалим

**Review by AlexMasterov — CHANGES_REQUESTED** [review]

File: src/main.js (hunk 9 above)
Comment (AlexMasterov) [inline src/main.js:12]: <details>
<summary>utils/utils.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/utils.js#L3
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/utils.js#L11-L21
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/utils.js#L32-L44
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/utils/utils.js#L46-L51
</details>

Для работы с датой создай отдельный модуль `utils/date.js`.

**Review by AlexMasterov — CHANGES_REQUESTED** [review]

File: src/presenter/films-board.js [hunk 10]
```diff
@@ -2,143 +2,187 @@ import FilmsListView from "./../view/films-list.js";
...
+    this._filterModel = filterModel;
+    this._showMoreButtonComponent = null;
     this._renderedFilmsCount = FILMS_COUNT_PER_STEP;
     this._currentSortType = SortType.DEFAULT;
     this._filmsListComponent = new FilmsListView();
-    this._sortComponent = new SortingView();
+    this._sortComponent = null;
     this._filmCardPresenter = {};
     this._filmsContainerComponent = new FilmsContainerView();
     this._filmsListTitleComponent = createFilmListTitleComponent;
```
Comment (AlexMasterov) [inline src/presenter/films-board.js:27]: <details>
<summary>factories/film-list-title.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/factories/film-list-title.js#L1-L3
</details>

Модуль удали и пока снова создавай компонент через `new`.

**Review by AlexMasterov — CHANGES_REQUESTED** [review]

File: src/view/film-card-full.js [hunk 11]
```diff
@@ -1,11 +1,11 @@
+import he from "he";
 import {TIME_MARKS} from "../utils/const.js";
-import {humanizeDuration} from "../utils/utils.js";
-import AbstractView from "./abstract.js";
```
Comment (AlexMasterov) [inline src/view/film-card-full.js:3]: <details>
<summary>view/abstract.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/view/abstract.js#L18
</details>

```diff
- if (!this._element) {
+ if (this._element === null) { 
```

File: src/main.js (hunk 9 above)
Comment (AlexMasterov) [inline src/main.js:12]: <details>
<summary>view/film-card-full.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/view/film-card-full.js#L10-L23
</details>

Перенеси в `utils/date.js`.

**Review by AlexMasterov — CHANGES_REQUESTED** [review]

File: src/main.js (hunk 9 above)
Comment (AlexMasterov) [inline src/main.js:12]: <details>
<summary>view/film-card.js</summary>

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/view/film-card.js#L98
https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/448bc73532010696a8e058ea63e02c7ce65e34eb/src/view/film-card.js#L102
</details>

.film-card__controls-item--`mark-as-watched` !== this._`watched`ClickHandler
.film-card__controls-item--`add-to-watchlist` !== this._`watching`ClickHandler

Д4

**Review by AlexMasterov — CHANGES_REQUESTED** [review]

File: src/presenter/films-board.js (hunk 3 above)
Comment (SampetovaN) [reply src/presenter/films-board.js:51]: да, вот с этим какая-то проблема вышла. если убираю slice, то новый массив для дефолтного списка не возращается. долго сидела, но так и не доехала, почему список не обновляется... подумала, что проблема в функции sortFilms

File: src/view/film-card-full.js [hunk 12]
```diff
@@ -182,56 +199,126 @@ const createFilmCardFullTemplate = (filmCard, commentData) => {
...
+    this._callback.deleteClick = callback;
+    this.element.addEventListener(`click`, this._commentDeleteClickHandler);
+  }
+  _renderNewComment(textValue) {
+    return {
+      date: humanizeCommentDateTime(new Date()),
+      emoji: this._data.emojiValue.value,
+      author: COMMENT_AUTHORS[0],
+      text: he.encode(textValue),
+      id: nanoid(),
```
[the remark this answers is not in the document]
Comment (SampetovaN) [reply src/view/film-card-full.js:303]: так если я сейчас удалю id, то я не смогу удалить комментарий. я же нахожу комментарий по id и его удаляю. Подскажи, пожалуйста, куда мне надо перенести временно формирование id?

File: src/presenter/films-board.js (hunk 3 above)
Comment (AlexMasterov) [reply src/presenter/films-board.js]: > если убираю `slice`, то новый массив для **дефолтного списка** (= `FilterType.ALL`) не возращается

https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/7bba473303092918ce176cd235fe0dd5a50c23a1/src/utils/filter.js#L4

Обычно от функции фильтрации (наивно) ожидают, что новый массив вернется в любом случае, но здесь возвращается старый 😃

**Review by AlexMasterov — APPROVED** [review]

**Comment by AlexMasterov** [comment]
Пожалуйста, посмотри чат и далее делай задание из раздела "Работа с сетью", чтобы естественным образом появилась долгожданная структура с комментариями 😃

На этом задание нет смысла задерживаться.

File: src/view/film-card-full.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (AlexMasterov) [reply src/view/film-card-full.js:300]: Например, мне пришлось временно добавить модели `films` метод [`_deleteComment`](https://github.com/htmlacademy-ecmascript/1236041-cinemaddict-12/blob/7bba473303092918ce176cd235fe0dd5a50c23a1/src/view/film-card-full.js#L269-L281).

Не советую 😄
Предлагаю скорее реализовать получения данных с сервера.

**Pull request merged**